packages/astro/src/core/app/node.ts

```diff
@@ -36,18 +36,27 @@ export class NodeApp extends App {
 	render(req: IncomingMessage | Request, routeData?: RouteData) {
 		if ('on' in req) {
 			let body = Buffer.from([]);
-			let reqBodyComplete = new Promise((resolve, reject) => {
-				req.on('data', (d) => {
-					body = Buffer.concat([body, d]);
+			let reqBodyComplete = null;
+			// @ts-ignore
+      if (req.body) {
+        reqBodyComplete = new Promise((resolve, reject) => {
+					// @ts-ignore
+					body = Buffer.from(JSON.stringify(req.body));
```

Comment:
I'm guessing this is what is causing the tests to fail. req.body is not a string here, I imagine.

Reply:
The Express JSON middleware is used by several other Express middleware and it parses the POST request data into the req.body. This is how I fixed it to allow the Astro POST end-points to work. Maybe it is not the best way to fix this issue or I could try changing the if statement. Is there there a way for me to just run the unit test that are failing locally?

Reply:
I have figure how to run the node unit test locally, so will work on a better fix.

Reply:
I changed the if statement to:
if ((typeof(req.body) === 'object') && (Object.keys(req.body).length > 0)) {

Reply:
Are you expecting `req.body` to be a string here? If so just check that directly: `if(req.body === 'string')`...

Reply:
With these standard Express middleware being used when a POST request is being processed the data is parsed into the req.body property as an object with properties.
app.use(express.json()) // for parsing application/json
app.use(express.urlencoded({ extended: true })) // for parsing application/x-www-form-urlencoded

Reply:
Ah, I understand now. In that case might I suggest using something like this: https://github.com/sindresorhus/is-plain-obj . That will probably get around the problem you are running into where the body is an object but not one you can stringify.

Reply:
I am doing the test for properties "(Object.keys(req.body).length > 0)", because found in the unit testing that req.body is being set to an empty object and causing the unit test to fail. I am not seeing that is-plain-obj has an option to test if an object is plain and not empty.

Reply:
Where does the unit test do that?

Reply:
@matthewp I am not familiar with JavaScript unit testing, I am more with Java, I just know when I was debugging while running the test what I was seeing. I will try to find the JavaScript unit test and pin point what was causing the test to fail. Also I have never worked with a JavaScript project of this magnitude and do not know how to install a dependency for a particular module. Any help in these areas would be appreciated.

Reply:
@matthewp I am making some progress with the unit test. Found that the unit test packages\integrations\node\test\api-route.test.js on line 20 "let { req, res, done } = createRequestAndResponse({" it calls createRequestAndResponse from import of packages\integrations\node\test\test-utils.js. Then in createRequestAndResponse on line 20 "let req = httpMocks.createRequest(reqOptions);" it calls httpMocks.createRequest from import of node-mocks-http. Looking at the source for [httpMocks.createRequest](https://github.com/howardabrams/node-mocks-http/blob/master/lib/mockRequest.js) on line 73 "mockRequest.body = options.body ? options.body : {};" the Request that is created will always have a body property that is at least an empty object. I know this is a lot to share, but hopefully helps us.